tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs

```diff
@@ -17,5 +18,104 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
+        {
+            switch (action.CommandId)
+            {
+                case "createCard":
+                    return HandleCreateCardCommands(turnContext, action);
+
+                case "shareMessage":
+                    return HandleShareMessageCommand(turnContext, action);
+                default:
+                    return new MessagingExtensionActionResponse();
+            }
+        }
+
+        private MessagingExtensionActionResponse HandleCreateCardCommands(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
+        {
+            var actionData = (action.Data as JObject).ToObject<CreateCardData>();
+            var response = new MessagingExtensionActionResponse
+            {
+                ComposeExtension = new MessagingExtensionResult
+                {
+                    AttachmentLayout = "list",
+                    Type = "result",
+                },
+            };
+
+            var card = new HeroCard
+            {
+                Title = actionData.Title,
+                Subtitle = actionData.Subtitle,
+                Text = actionData.Text,
+            };
+
+            var attachments = new List<MessagingExtensionAttachment>();
+            attachments.Add(new MessagingExtensionAttachment
+            {
+                Content = card,
+                ContentType = HeroCard.ContentType,
+                Preview = card.ToAttachment(),
+            });
+
+            response.ComposeExtension.Attachments = attachments;
+            return response;
+        }
+
+        private MessagingExtensionActionResponse HandleShareMessageCommand(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
+        {
+            var heroCard = new HeroCard
+            {
+                Title = $"{action.MessagePayload.From.User.DisplayName} sent this message:",
+                Text = action.MessagePayload.Body.Content,
+            };
+            var includeImageData = (action.Data as JObject).ToObject<IncludeImageData>();
+            if (includeImageData.IncludeImage)
+            {
+                heroCard.Images = new List<CardImage>
+                {
+                    new CardImage
+                    {
+                        Url = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU",
+                    },
+                };
+            }
+
+            var response = new MessagingExtensionActionResponse();
+            response.ComposeExtension = new MessagingExtensionResult
+            {
+                Type = "result",
+                AttachmentLayout = "list",
+                Attachments = new List<MessagingExtensionAttachment>
+                {
+                    new MessagingExtensionAttachment
```

Comment:
This is slightly ridiculous. At the very least a ToMessagingExtensionAttachment(). But ideally we could do better than that.

Reply:
Resolving as issue #2603 was filed to address this feedback.